### ISSUE TITLE
Backend support in 2.0

### ISSUE THREAD

**Issue by bgroenks96** [nifty-gui/nifty-gui#334]
@void256 @3xp0n3nt 

Should Nifty 2.0 support all the different backends the way it does now? Or should this be exported to the CoreGL project?

**Comment by void256** [comment]
Right now Nifty 2.0 only has [NiftyRenderDeviceOpenGL](https://github.com/void256/nifty-gui/blob/2.0/nifty-renderer-opengl/src/main/java/de/lessvoid/nifty/renderer/opengl/NiftyRenderDeviceOpenGL.java) which uses brute force CoreGL as it's render backend.

This allows us to have **ONE** NiftyRenderDevice implementation and still support LWJGL **AND** JOGL as the actual GL adapter (this is one of the main tasks of the CoreGL project anyway). From my point of view this is a big plus because we only have to implement the GL stuff once and let CoreGL handle the "translation" to LWJGL and JOGL.

That being said, there is no real reason to not have additional NiftyRenderDevice implementations for other graphics API (Java2D or Slick or Metal or Vulkan ... :wink:) but right now I, for one, are concentrating on getting Nifty 2.0 into a somewhat usable state so other RenderDevices are very low priority on my list.

Which doesn't mean I'm against it in any way btw :smile:

**Comment by bgroenks96** [comment]
Right, but would it be feasible to work on building support for those graphics APIs into CoreGL? That way Nifty (and other projects) could hypothetically support them all while having only one OpenGL interface to maintain?

**Comment by void256** [comment]
CoreGL is very OpenGL specific. I'm not sure if it makes a lot of sense to integrate for instance Java2D as a CoreGL backend o_O If you mean that kind of things :wink: A CoreShader or CoreVBO is all very much specific to the API that OpenGL offers .. emulating it with other libs might get cumbersome or impossible.

**Comment by ghost** [comment]
If anyone is still interested (I know it's not a priority for Nifty), I'd be interested in porting an upgraded (current one has code rot) LibGDX backend into 2.0, but it would require the Nifty rendering system abstraction layer to be modular enough that the parts of it which require OpenGL functionality that is non-existent in OpenGL ES / LibGDX (such as geometry shaders) could be re-written to work without it. (By re-written, I mean providing alternate "pluggable" implementations).

The easy part is creating an ESGL interface as a subset of a CoreGL interface. The hard part (which I would be willing to do if designed in such a way as to make it feasible) is to write alternate implementations of certain rendering API's in Nifty that don't use non-existent OpenGL functionality, but provide an end result that is virtually indistinguishable visually from the original (albeit with less performance). As an analogy, it's similar to what @bgroenks96 
did with the Java2D rendering backend in 1.4.

The other difficulty would be not allowing callers direct access to certain advanced OpenGL features that are not available in ES. Meaning that not only does Nifty use these features internally with CoreGL, but provides pass-through access to callers who want to directly play with OpenGL.

**Comment by bgroenks96** [comment]
@void256 I didn't mean stuff like Java2D. That would require separate implementations in Nifty. Admittedly, however, the same problem probably applies to stuff like Slick and LibGDX because they are too high level to easily be written to implement the CoreGL interface.

@3xp0n3nt That's probably a good idea. Hopefully though yours would work, unlike my Java2D backend (it's still completely broken at the moment).

**Comment by mkaring** [comment]
I guess the people using the jMonkeyEngine will need a dedicated backend
that ties into the engine as well. So I think in the end it has to work
without coregl.

Brian Groenke notifications@github.com schrieb am Mi., 29. Apr. 2015
00:49:

**Comment by bgroenks96** [comment]
Yeah it just kind of sucks that all of those backends have to be reworked again. How much of the current RenderDevice spi is being kept?

**Comment by ghost** [comment]
It would be nice if OpenGL support could be pluggable, meaning abstracted enough in Nifty that you could use any type of rendering, even DirectX.

**Comment by bgroenks96** [comment]
@3xp0n3nt Ewww. Don't you dare say the D word around here :P

**Comment by ghost** [comment]
Just an example to think outside the OpenGL box. Can we remove all OpenGL-specific assumptions from Nifty's rendering abstraction? Or maybe that's already done in 2.0, I haven't looked at it lately.

**Comment by void256** [comment]
Ok!

I think we need to clarify some things:

1) We still have the basic [NiftyRenderDevice](https://github.com/void256/nifty-gui/blob/2.0/nifty/src/main/java/de/lessvoid/nifty/spi/NiftyRenderDevice.java) SPI available in 2.0! This can always be implemented for whatever new rendering system needed. Even DirectX :wink: This interface is abstracted enough that it is not dependent on GL at all.

2) The NiftyRenderDevice for 2.0 is a little bit more complicated than the one in 1.x. There are the following three more advanced features required:
- **render to texture** - should be easy for implementations
- **some path rendering** - f.i. lines and arcs and still to be done filled path segments (urgs, this might be a bit ugly)
- **custom shaders** - however from the point of view of the NiftyRenderDevice these are just resources that can be loaded and activated. We don't expect a GLSL emulation or something :wink: If there ever be a DirectX RenderDevice and you activate a custom shader then it should load and use a DirectX shader version of course!

3) Currently we only have one implementation available that is called [NiftyRenderDeviceOpenGL](https://github.com/void256/nifty-gui/blob/2.0/nifty-renderer-opengl/src/main/java/de/lessvoid/nifty/renderer/opengl/NiftyRenderDeviceOpenGL.java) which currently uses the CoreGL lib - which right now supports Core Profile OpenGL only using LWJGL or JOGL. I don't see any reason why CoreGL could not be extended use OpenGL ES too. Eventually emulating some features as @3xp0n3nt said. But this only makes sense when it fits into the general API of the CoreGL project. If its too messy then I still think a dedicated NiftyRenderDeviceOpenGLES would be better but we should decide that after someone took a closer look, I think.

**Comment by ghost** [comment]
@void256 Thanks! That really clears everything up for me.

**Comment by spaceemotion** [comment]
If 2.0 should support a Java2D backend, the custom shaders could be a problem. Other than that, it should work perfectly fine. I actually planned on writing a custom implementation for my UI editor (one that supports version 2.0) inside ready to be used inside swing applications. For Java2D I would be perfectly fine with dropping shader support but I don't know how your system would handle this, since you seem to use shaders for the simplest things even…

**Comment by void256** [comment]
The current OpenGL implementation of the NiftyRenderDevice for Nifty 2.0 uses shaders for everything, yes! In the 21st century that's the way things are supposed to work anyway :wink:

A Java2D implementation of the NiftyRenderDevice doesn't have to use shaders at all! So there really is no problem with Java2D concerning shaders!

Besides one tiny exception:

Nifty 2.0 supports custom shaders which means you can render something using user provided shader code. That feature will not be available in a Java2D implementation of course. But if you target Java2D you really should be aware that you can't use that custom shader feature and shouldn't be surprised about missing support for that :smile:

So to summarize: Java2D support should be fine!

**Comment by spaceemotion** [comment]
@void256 awesome!

Guess you can sign me up for a possible Java2D implementation then :)
I currently don't know when I'm ready for writing that stuff though, but definitely this year.

**Comment by bgroenks96** [comment]
@void256 @spaceemotion I already have a partially working Java2D implementation working for Nifty 1.4. Will that work in 2.0 or did the interface change?

Granted, it's currently broken because of weird coordinate translation issues (and I would appreciate help fixing it up), but nonetheless, there's something there.

**Comment by void256** [comment]
The interface changed in 2.0 so you can't use it directly. However most of it still works similar so it shouldn't be too hard to adapt it to 2.0. You might need a way to render stuff into a backbuffer and later draw that buffer but I guess you can do that?

I'd be interested how Java2D performs in 2.0! So yes, please go ahead and integrate it, if you can :smile: Should I prepare a sub module for it so that you have a place to start?

**Comment by bgroenks96** [comment]
> You might need a way to render stuff into a backbuffer and later draw that buffer but I guess you can do that?

Yes you can. The Java2D BatchRenderBackend implementation does that currently.

> Should I prepare a sub module for it so that you have a place to start?

Sure! I'll let you know once I get it working properly in 1.4.

**Comment by void256** [comment]
I've prepared a NiftyRenderDeviceJava2D and added support for it to the UseCaseRunner in the examples project, see this commit for details: 2a343b38fca322336fea063e943811f11de9708f

**Comment by bgroenks96** [comment]
Since I honestly don't really want to continue down the awful Java2D porting road... I would like to propose an alternative that might solve the issues of those who want Java2D support like @relu91 

JOGL supports integration with Swing/AWT via their [NewtCanvasAWT](http://jogamp.org/deployment/jogamp-next/javadoc/jogl/javadoc/com/jogamp/newt/awt/NewtCanvasAWT.html). I would prefer that we encourage Swing/AWT users to migrate to this instead of requiring a port of the increasingly outdated and unworkable Java2D rendering system.

@relu91 I would be willing to help you migrate NiftyEditor to use the JOGL backend instead of Java2D and see if this is a viable solution. We can even see along the way what steps can be taken to add features to Nifty (or CoreGL) and make the process even easier for the end user.

**Comment by bgroenks96** [comment]
If I may, let me go ahead and tag @gouessej in here as well. Does it sound viable to replace the Nifty Java2D backend (or at least its purpose) with just JOGL and NewtCanvasAWT, Julien? Are there any limitations we should be aware of that might warrant keeping pure Java2D support around?

**Comment by ghost** [comment]
In my humble opinion, dropping an half broken Java2D backend to concentrate on the JogAmp backend is a good idea. If the Java2D backend worked very well, it would be worth to keep it. The JOGL AWT GLCanvas is still more reliable than NewtCanvasAWT (which is still faster) and GLJPanel solves a few problems with transparent components. Personally, I would support them all like JogAmp's Ardor3D Continuation (which supports NewtCanvasSWT and SWT GLCanvas too). As a first step, maybe consider only GLWindow and NewtCanvasAWT and keep other options if and only if the other end users are concerned by the limitations or bugs of NewtCanvasAWT. There are a few remaining problems in the focus handling and sometimes in the event mapping with AWT.

**Comment by relu91** [comment]
I'm going to rewrite the editor from scratch for the nifty 2.0 version, so it's ok to use an other renderer! Also I hope to use nifty-gui even for the gui editor it-self,which will be very cool in my opinion. Thus we could even avoid a java2d/awt integration! :smile:

**Comment by mkaring** [comment]
I agree as well. Dropping Java2D for Nifty 2.0 is more viable then trying to get it working to some extend. I think the JOGL canvas things should already work, since they are also "just" a OpenGL context. Shouldn't CoreGL cover that? At least mostly?

@relu91 I think it would be great if you could look at the code of Nifty 2.0 from time to time from the perspective of writing the editor. Especially regarding things like discovering the required information for the editor. Currently we "only" have the basic nodes. But those are already a very important part for the editor.

**Comment by relu91** [comment]
@mkaring yes I'm already doing that! What I find little difficult right now is to understand the new layout nodes is there a pull request where it was explained the new architetture? 

OT: It would be also awesome if we can have a tool to see the development process, I like the github issue system but I get lost easly. Could we make a standard tag for "developing issues" ? I don't know f.I. 

Issue title: Add again the controls with the new layout system
Description: Now that we have a new layout system the control should be like this. < Insert some documentation about the new architetture that needs to be developed>
tag:Nifty 2.0 , Development

I don't know if it's actually doable but I miss something like this to see the big picture. It would also help others to contribute. 
What do you think @void256 ?

**Comment by mkaring** [comment]
@relu91 I do not think there is, I am currently in progress of adding a lot of JavaDoc documentation to the layout classes trying to explain how they work and what the basic ideas are. If you want we can also chat regarding the layout system. I am very interested in your opinion, but a discussion like this is OT for this issue.

**Comment by bgroenks96** [comment]
@gouessej Thanks. That sounds good.

@mkaring Yes, CoreGL _should_ already work with this provided that the user is using the JogAmp backend and does all of the setup themselves. However, since the Java2D backend sort of existed for the purpose of being an easy alternative for new users or those with basic needs, I'm suggesting we could take care of some of this boilerplate for them, in Nifty; i.e. we can set up the rendering context and just return a GLJPanel or NewtCanvasAwt that they can treat as any other Java2D component.

This could be done in the existing Java2D renderer backend module or in the JogAmp one. Doesn't matter too much.

Obviously it will require JogAmp as a transient dependency, but the end-user will pretty much be able to focus on just Swing/AWT and won't even have to be familiar with JOGL to use it.

**Comment by mkaring** [comment]
So we are talking a `NiftyJPanel` or a `NiftyCanvas` that does… things and makes everything working. But I am not sure if we should declare a utility classes like this as Java2D render module or rather as Java2D utility module that uses the OpenGL renderer of Nifty along with JOGL and their AWT/Swing classes.

Just to avoid duplicated code. ;)

**Comment by bgroenks96** [comment]
We could make a new module called `nifty-renderer-jogl-awt`  .... or maybe `nifty-renderer-awt-opengl`

Calling it Java2D would be misleading since it wouldn't be using the Java2D API... just integration with Swing/AWT.
